code/game/gamemodes/gang/vigilante.dm

```diff
@@ -0,0 +1,215 @@
+		/datum/gang_item/equipment/shield,
+		/datum/gang_item/equipment/bulletproof_armor,
+		/datum/gang_item/equipment/bulletproof_helmet,
+		/datum/gang_item/equipment/gangbreaker,
+		/datum/gang_item/equipment/seraph
+		)
+
+/obj/item/device/vigilante_tool/Initialize()
+	if(ismob(loc))
+		vig_item_list = list()
+		vig_category_list = list()
+		for(var/V in vigilante_items)
+			var/datum/gang_item/G = new V()
+			vig_item_list[G.id] = G
+			var/list/Cat = vig_category_list[G.category]
+			if(Cat)
+				Cat += G
+			else
+				vig_category_list[G.category] = list(G)
+		var/mob/living/M = loc
+		linked_action = new(src)
+		linked_action.Grant(M, src)
+		addtimer(CALLBACK(src, .proc/earnings), 1500)
+
+/obj/item/device/vigilante_tool/Destroy()
+	var/mob/living/M = loc
+	linked_action.Remove(M)
+	M.update_icons()
+	return ..()
+
+/obj/item/device/vigilante_tool/proc/earnings()
+	if(!src)
+		return
+	var/all_territory = list()
+	var/newpoints = 0
+	var/mob/living/carbon/human/H = loc
+	for(var/datum/gang/G in SSticker.mode.gangs)
+		all_territory += G.territory
+	for(var/area/A in tags)
+		if(!(A in all_territory))
+			newpoints += 0.5
+	to_chat(H, "<span class='notice'>You have received 4 influence for your continued loyalty, [newpoints] for keeping the station tag-free.")
+	points += newpoints + 4
+	for(var/obj/item/weapon/implant/mindshield/I in H.implants)
+		points += 4
+		to_chat(H, "<span class='notice'>You have also received 4 influence for possessing a mindshield implant.</span>")
+	addtimer(CALLBACK(src, .proc/earnings), 1500, TIMER_UNIQUE)
+
+/obj/item/device/vigilante_tool/attack_self(mob/user)
+	if(user.mind in SSticker.mode.get_all_gangsters())
+		return
+	var/dat
+	dat += "Your Influence: <B>[points]</B><br>"
+	dat += "<center><a href='?src=\ref[src];destroy=TRUE'><B>DESTROY HELD CONTRABAND</a></center></B><br>"
+	dat += "<hr>"
+	for(var/cat in vig_category_list)
+		dat += "<b>[cat]</b><br>"
+		for(var/V in vig_category_list[cat])
+			var/datum/gang_item/G = V
+			var/cost = G.get_cost_display(user, gang, src)
+			if(cost)
+				dat += cost + " "
+			var/toAdd = G.get_name_display(user, gang, src)
+			if(G.can_buy(user, gang, src))
+				toAdd = "<a href='?src=\ref[src];purchase=[G.id]'>[toAdd]</a>"
+			dat += toAdd
+			var/extra = G.get_extra_info(user, gang, src)
+			if(extra)
+				dat += "<br><i>[extra]</i>"
+			dat += "<br>"
+		dat += "<br>"
+
+	dat += "<a href='?src=\ref[src];choice=refresh'>Refresh</a><br>"
+
+	var/datum/browser/popup = new(user, "gangtool", "Welcome to Vigilante's Companion v1.1", 400, 750)
+	popup.set_content(dat)
+	popup.open()
+
+
+
+/obj/item/device/vigilante_tool/Topic(href, href_list)
+	if(usr.restrained() || usr.lying || usr.stat || usr.stunned || usr.weakened)
+		return 0
+	if(href_list["purchase"])
+		var/datum/gang_item/G = vig_item_list[href_list["purchase"]]
+		if(G && G.can_buy(usr, gang, src))
+			G.purchase(usr, gang, src, FALSE)
+	if(href_list["destroy"])
+		Destroy_Contraband(usr)
+	attack_self(usr)
```

Comment:
don't call attack_self, refactor what's in attack_self into a unique proc and call that